it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java

```diff
@@ -0,0 +1,251 @@
+    this.topicAdminClient = topicAdminClient;
+    this.subscriptionAdminClient = subscriptionAdminClient;
+    this.createdTopics = Collections.synchronizedSet(new HashSet<>());
+    this.createdSubscriptions = Collections.synchronizedSet(new HashSet<>());
+  }
+
+  public static Builder builder(String testName, String projectId) {
+    checkArgument(!Strings.isNullOrEmpty(testName), "testName can not be null or empty");
+    checkArgument(!projectId.isEmpty(), "projectId can not be empty");
+    return new Builder(testName, projectId);
+  }
+
+  /**
+   * Create a topic based on the given topic name. However, it adds a prefix based on the test id
+   * and timestamp, to uniquely identify the current execution instance.
+   *
+   * @see PubsubResourceManager#createTopic(String)
+   */
+  @Override
+  public TopicName createTopic(String topicName) {
+    checkArgument(!topicName.isEmpty(), "topicName can not be empty");
+    checkIsUsable();
+
+    LOG.info("Creating topic '{}'...", topicName);
+
+    Topic topic = topicAdminClient.createTopic(getTopicName(topicName));
+    TopicName reference = PubsubUtils.toTopicName(topic);
+    createdTopics.add(reference);
+
+    LOG.info("Topic '{}' was created successfully!", topicName);
+
+    return reference;
+  }
+
+  /**
+   * Create a subscription based on the given name. However, it adds a prefix based on the test id
+   * and timestamp, to uniquely identify the current execution instance.
+   *
+   * @see PubsubResourceManager#createSubscription(TopicName, String)
+   */
+  @Override
+  public SubscriptionName createSubscription(TopicName topicName, String subscriptionName) {
+    checkArgument(!subscriptionName.isEmpty(), "subscriptionName can not be empty");
+    checkIsUsable();
+
+    if (!createdTopics.contains(topicName)) {
+      throw new IllegalArgumentException(
+          "Can not create a subscription for a topic not managed by this instance.");
+    }
+
+    LOG.info("Creating subscription '{}' for topic '{}'", subscriptionName, topicName);
+
+    Subscription subscription =
+        subscriptionAdminClient.createSubscription(
+            getSubscriptionName(subscriptionName),
+            topicName,
+            PushConfig.getDefaultInstance(),
+            DEFAULT_ACK_DEADLINE_SECONDS);
+    SubscriptionName reference = PubsubUtils.toSubscriptionName(subscription);
+    createdSubscriptions.add(reference);
+
+    LOG.info(
+        "Subscription '{}' for topic '{}' was created successfully!", topicName, subscriptionName);
+
+    return reference;
+  }
+
+  @Override
+  public String publish(TopicName topic, Map<String, String> attributes, ByteString data)
+      throws PubsubResourceManagerException {
+    checkIsUsable();
+
+    LOG.info("Publishing message with {} bytes to topic '{}'", data.size(), topic);
+
+    PubsubMessage pubsubMessage =
+        PubsubMessage.newBuilder().putAllAttributes(attributes).setData(data).build();
+
+    try {
```

Comment:
Forgot to mention this last time: We should verify that the topic being published to is tracked by the resource manager. We don't want tests to write to any topic in the project.